### ISSUE TITLE
rclone copy creating empty directories it didn't create before

### ISSUE THREAD

**Issue by YukiUnHappy** [rclone/rclone#7689]
#### What is the problem you are having with rclone?

`Rclone will now sync directory modification times if the backend supports it`
That's a cool new feature, but I think it shouldn't be enabled by default, or it should ignore folders by default where no files need to be transferred.
In the previous version, I could simply `copy` to the remote within the folder where I needed to upload files, but now I have to remove the other folders in that folder that have no files first, otherwise these empty folders will also be created on the remote due to the sync modification time.
Conversely, if I only need to copy specific files from the remote to the local (using `--include`), but other irrelevant folders under that path on the remote will also be created on my local machine with the original structure. This did not exist in previous versions and is a bit annoying.

#### What is your rclone version (output from `rclone version`)

rclone v1.66.0
- os/version: debian 10.0 (64 bit)
- os/kernel: 5.14.9 (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.22.1
- go/linking: static
- go/tags: none

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)

The information has already been mentioned above.

#### Which cloud storage system are you using? (e.g. Google Drive)

This is not related to a specific backend.

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)

`rclone copy -vv Source Remote --include E.txt`

#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)

```shell
root@localhost:~/TEST# tree
.
└── Source
    ├── A
    │   └── D
    ├── B
    ├── C
    └── E.txt

5 directories, 1 file
root@localhost:~/TEST# rclone copy -vv Source Remote --include E.txt
2024/03/24 23:51:52 DEBUG : rclone: Version "v1.66.0" starting with parameters ["rclone" "copy" "-vv" "Source" "Remote" "--include" "E.txt"]
2024/03/24 23:51:52 DEBUG : Creating backend with remote "Source"
2024/03/24 23:51:52 DEBUG : Using config file from "/root/.config/rclone/rclone.conf"
2024/03/24 23:51:52 DEBUG : fs cache: renaming cache item "Source" to be canonical "/root/TEST/Source"
2024/03/24 23:51:52 DEBUG : Creating backend with remote "Remote"
2024/03/24 23:51:52 DEBUG : fs cache: renaming cache item "Remote" to be canonical "/root/TEST/Remote"
2024/03/24 23:51:52 DEBUG : A: Making directory with metadata
2024/03/24 23:51:52 INFO  : A: Made directory with metadata (mtime=2024-03-24T23:50:07.378243503-07:00)
2024/03/24 23:51:52 DEBUG : Added delayed dir = "A", newDst=A
2024/03/24 23:51:52 DEBUG : B: Making directory with metadata
2024/03/24 23:51:52 INFO  : B: Made directory with metadata (mtime=2024-03-24T23:50:01.04917523-07:00)
2024/03/24 23:51:52 DEBUG : Added delayed dir = "B", newDst=B
2024/03/24 23:51:52 DEBUG : C: Making directory with metadata
2024/03/24 23:51:52 INFO  : C: Made directory with metadata (mtime=2024-03-24T23:50:03.246198928-07:00)
2024/03/24 23:51:52 DEBUG : Added delayed dir = "C", newDst=C
2024/03/24 23:51:52 DEBUG : E.txt: Need to transfer - File not found at Destination
2024/03/24 23:51:52 DEBUG : A/D: Making directory with metadata
2024/03/24 23:51:52 INFO  : A/D: Made directory with metadata (mtime=2024-03-24T23:50:07.378243503-07:00)
2024/03/24 23:51:52 DEBUG : Added delayed dir = "A/D", newDst=A/D
2024/03/24 23:51:52 DEBUG : Local file system at /root/TEST/Remote: Waiting for checks to finish
2024/03/24 23:51:52 DEBUG : E.txt: md5 = d8e8fca2dc0f896fd7cb4cb0031ba249 OK
2024/03/24 23:51:52 DEBUG : E.txt.cihekil4.partial: renamed to: E.txt
2024/03/24 23:51:52 INFO  : E.txt: Copied (new)
2024/03/24 23:51:52 DEBUG : Local file system at /root/TEST/Remote: Waiting for transfers to finish
2024/03/24 23:51:52 INFO  : A/D: Set directory modification time (using SetModTime)
2024/03/24 23:51:52 INFO  : C: Set directory modification time (using SetModTime)
2024/03/24 23:51:52 INFO  : A: Set directory modification time (using SetModTime)
2024/03/24 23:51:52 INFO  : B: Set directory modification time (using SetModTime)
2024/03/24 23:51:52 INFO  :
Transferred:              5 B / 5 B, 100%, 0 B/s, ETA -
Transferred:            1 / 1, 100%
Elapsed time:         0.0s

2024/03/24 23:51:52 DEBUG : 7 go routines active
root@localhost:~/TEST# tree
.
├── Remote
│   ├── A
│   │   └── D
│   ├── B
│   ├── C
│   └── E.txt
└── Source
    ├── A
    │   └── D
    ├── B
    ├── C
    └── E.txt

10 directories, 2 files
```

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by sabitm** [comment]
I'm not 100% sure if this is related, but `bisync` also has a weird behavior on 1.66.0. It keep producing a file conflict (which I have to `resync` to resolve it) between local and remote storage. I just assume that is because of `experimental` nature of bisync, hence I'm not opening an issue and will happily use the previous version for time being.

**Comment by nielash** [comment]
Hi @sabitm, `bisync` maintainer here. `bisync` received a [very large update](https://rclone.org/bisync/#changelog) in `v1.66`, and [handling of sync conflicts](https://rclone.org/bisync/#conflict-resolve) is indeed one of the things that changed. So it may be that the new behavior you're seeing is intentional, but if there's a bug then I'd definitely like to know about it. Please feel free to open an issue or post in the [forum](https://forum.rclone.org) with more details!

As for the original issue -- FYI there is an ongoing discussion about it in the forum:
https://forum.rclone.org/t/how-to-ignore-empty-directories-when-uploading-from-windows/45057/9

**Comment by sabitm** [comment]
Hi @nielash, thanks for maintaining `bisync`!

I'm not entirely sure how to debug though. It happens randomly enough. I'll try to find a way to reproduce it and will opening an issue if I found something interesting.

**Comment by ncw** [comment]
There is quite a long thread on the forum about this

https://forum.rclone.org/t/empty-dirs-not-wanted/45059/

For rclone v1.66.1 I am going to make sure it obeys the `--create-empty-src-dirs` flag.

What the defaults for this for `rclone sync`, `rclone copy` and `rclone move` I have yet to decide. Currently they are all false. There is an argument for making this `true` for rclone sync but that might be a step too far for a point release.

**Comment by ncw** [comment]
@YukiUnHappy 
> Additionally, I have a small issue that I won't open a separate issue for. The auto completion improvement in 1.66 has broken the ability to complete paths with spaces in bash.

I can confirm that. Can you open a [new issue](https://github.com/rclone/rclone/issues/new/choose) about this please?  So we don't mix the two up - thank you. (BTW I think this is a bug in cobra and here is the fix https://github.com/spf13/cobra/pull/2126 - Note also that if you type a `"` first eg `"D<TAB>` you'll get the correct completions.)

**Comment by ncw** [comment]
Here is a first attempt at a fix - this obeys `--create-empty-src-dirs` in the same way that 1.65.2 does

[v1.67.0-beta.7854.8d02ad3ba.fix-7689-empty-dirs](https://beta.rclone.org/branch/fix-7689-empty-dirs/v1.67.0-beta.7854.8d02ad3ba.fix-7689-empty-dirs/) on branch [fix-7689-empty-dirs](https://github.com/rclone/rclone/tree/fix-7689-empty-dirs) (uploaded in 15-30 mins)

@nielash I've been unable to get the bisync tests to pass on Windows or macOS. I have no idea why they are failing. Running with -golden sorted out the linux tests but not Windows or macOS. Can you help? Also any code review gratefully received :-)

**Comment by nielash** [comment]
> @nielash I've been unable to get the bisync tests to pass on Windows or macOS. I have no idea why they are failing.

It looks to me like the issue is that `Size()` for the dir is returning an unexpected value. It is expecting `4Ki` but getting `96` on `mac_amd64`:

```
2024/04/05 15:31:30 | MISCOMPARE  -Golden vs +Results for  test.log
2024/04/05 15:31:30 | @@ -139 +139 @@
2024/04/05 15:31:30 | -NOTICE: subdir: Skipped update directory metadata as --dry-run is set (size 4Ki)
2024/04/05 15:31:30 | +NOTICE: subdir: Skipped update directory metadata as --dry-run is set (size 96)
2024/04/05 15:31:30 | @@ -141 +141 @@
2024/04/05 15:31:30 | -NOTICE: subdir: Skipped update directory metadata as --dry-run is set (size 4Ki)
2024/04/05 15:31:30 | +NOTICE: subdir: Skipped update directory metadata as --dry-run is set (size 96)
```

and `0` on windows:

```
2024/04/05 15:30:53 | MISCOMPARE  -Golden vs +Results for  test.log
2024/04/05 15:30:53 | @@ -139 +139 @@
2024/04/05 15:30:53 | -NOTICE: subdir: Skipped update directory metadata as --dry-run is set (size 4Ki)
2024/04/05 15:30:53 | +NOTICE: subdir: Skipped update directory metadata as --dry-run is set (size 0)
2024/04/05 15:30:53 | @@ -141 +141 @@
2024/04/05 15:30:53 | -NOTICE: subdir: Skipped update directory metadata as --dry-run is set (size 4Ki)
2024/04/05 15:30:53 | +NOTICE: subdir: Skipped update directory metadata as --dry-run is set (size 0)
```

This comment seems potentially relevant:

https://github.com/rclone/rclone/blob/05e5712bc48f9ac92a96739897783a4bb9da6977/backend/local/local.go#L1494-L1497

I can take a closer look later!

**Comment by nielash** [comment]
@ncw I'm doing the closer look at [fix-7689-empty-dirs](https://github.com/rclone/rclone/tree/fix-7689-empty-dirs) now and I think there's an issue with the `--create-empty-src-dirs=false` case. Here's a test that demonstrates it: https://github.com/nielash/rclone/commit/16d856ac195575bf7b4e78d28a0532f544fb9b59 

The same test would have passed previously.

**Comment by nielash** [comment]
@ncw added a few other inline comments:
https://github.com/rclone/rclone/commit/dd463b2b4ff1feb7c5aeb9569fbd8c105a6d18cb
https://github.com/rclone/rclone/commit/13c72a94efd3f3b97d451385b1e46b04a44063c5

and a test that (I think) catches another regression: https://github.com/nielash/rclone/commit/cd0891ddc23db6a99caf81061237a0882c98ced3

**Comment by ncw** [comment]
I finally got round to fixing up this code. Can you take a quick look @nielash ?

I think it is doing the right thing now. I left the fixes unsquashed so you can see them and I put your tests in (though I took one line out as I don't think it was correct any more).

[v1.67.0-beta.7901.7042810e8.fix-7689-empty-dirs](https://beta.rclone.org/branch/fix-7689-empty-dirs/v1.67.0-beta.7901.7042810e8.fix-7689-empty-dirs/) on branch [fix-7689-empty-dirs](https://github.com/rclone/rclone/tree/fix-7689-empty-dirs) (uploaded in 15-30 mins)

Thank you :-)

**Comment by ncw** [comment]
Thanks for the review @nielash 

This feels like the never ending bug!

I've fixed the issue that you pointed out (I realize I misunderstood the purpose of `s.modifiedDirs` it is to keep the modified status of the directories themselves, not the parents like we need for delayed modtime setting).

Please take another look!

I also fixed a bug pointed out in the forum https://forum.rclone.org/t/empty-dirs-not-wanted/45059/14

Hopefully if this checks out I'll squash it down, neaten it up and then that is the last bit needed for v1.66.1

[v1.67.0-beta.7906.91d78fd38.fix-7689-empty-dirs](https://beta.rclone.org/branch/fix-7689-empty-dirs/v1.67.0-beta.7906.91d78fd38.fix-7689-empty-dirs/) on branch [fix-7689-empty-dirs](https://github.com/rclone/rclone/tree/fix-7689-empty-dirs) (uploaded in 15-30 mins)

**Comment by nielash** [comment]
Added a comment for you inline. This is a tricky one indeed!

**Comment by ncw** [comment]
@nielash I can't find your comment! I thought I saw a notification but it has gone! Please post a link - thank you :-)

**Comment by nielash** [comment]
@ncw here you go! https://github.com/rclone/rclone/commit/91d78fd381d2678baf7e41b3a3d3a1fc1ecaddd6#r141149941

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.66.1 

Thanks for all your help @nielash :-)
